### PR TITLE
DataParallel: GPU imbalance warning

### PR DESCRIPTION
**Pull request by lemairecarl**
I have 2 GPUs in my machine: a TITAN X with 12 GB of memory, and a GT710 with 1 GB, that is used for X server. Tensorflow automatically discards the GT710, based on the number of compute cores, if I recall correctly.

I should have simply set `CUDA_VISIBLE_DEVICES` so that the GT710 is ignored. But I did not do so, thus when trying code that uses DataParallel, PyTorch wants to use the 1 GB GPU along with the 12 GB one.

This gets me "out of memory" errors, unless I drastically reduce the batch size. Then, it begins training, but it sometimes randomly crashes with a "illegal memory access" error (though this may have been fixed in 0.3.1). The latter error might be related to the fact that the GT710 has compute capability 3.5 while the TITAN X has 5.2.

For 1-2 months I thought that this code was buggy or incompatible with the last version of PyTorch. It's only yesterday, by looking at `nvidia-smi`, that I saw that data was sent to the GT710. I can be blamed for not having hidden the GT710. But I think this case should be handled better than by "out of memory" errors. I think this could save headaches to many folks that use enterprise/university hardware such as mine.

Maybe we could do similar checking than what Tensorflow does. Or maybe we could do something like this (see proposed changes). Sadly, this won't work; `max_memory_allocated()` does not return the memory capacity of the GPU. Is there a method that does that?

### CODE REVIEW

File: torch/nn/parallel/data_parallel.py [hunk 1]
```diff
@@ -54,6 +54,13 @@ def __init__(self, module, device_ids=None, output_device=None, dim=0):
 
         if device_ids is None:
             device_ids = list(range(torch.cuda.device_count()))
+
+        devices_memory = [torch.cuda.max_memory_allocated(i) for i in device_ids]
+        if max(devices_memory) / min(devices_memory) >= 3.0:
```
Comment (soumith) [inline torch/nn/parallel/data_parallel.py]: this is great, but I think we should universally warn on such detection. Can you move this to torch/cuda/__init__.py where the other warnings are about old GPUs

Comment (lemairecarl) [reply torch/nn/parallel/data_parallel.py]: I'd gladly do so, but `max_memory_allocated()` returns 0 on my machine for all GPUs. The name of the method is misleading.

Comment (soumith) [reply torch/nn/parallel/data_parallel.py]: what you want is not `max_memory_allocated`, you want the total device memory.
It looks like we didn't expose this in pytorch yet, let me send a PR in a few minutes

Comment (soumith) [reply torch/nn/parallel/data_parallel.py]: @lemairecarl see https://github.com/pytorch/pytorch/pull/5386

Comment (lemairecarl) [reply torch/nn/parallel/data_parallel.py]: awesome, thanks!

@apaszke suggested in #5386 that we expose the whole struct. Have all info, should we also check the number of cuda cores? The GT710 has 192 cores, while the TITAN X has 3584 (almost 20x).

**Comment by lemairecarl** [comment]
I moved my check in `cuda.__init__.py`, as suggested.

But I think we should also check the number of CUDA cores. For that, having the whole into struct as suggested in #5386 would be best.

File: torch/cuda/__init__.py [hunk 2]
```diff
@@ -103,18 +103,30 @@ def _check_capability():
...
-        name = get_device_name(d)
         if CUDA_VERSION < 8000 and major >= 6:
             warnings.warn(incorrect_binary_warn % (d, name, 8000, CUDA_VERSION))
         elif CUDA_VERSION < 9000 and major >= 7:
             warnings.warn(incorrect_binary_warn % (d, name, 9000, CUDA_VERSION))
         elif capability == (3, 0) or capability == (5, 0) or major < 3:
             warnings.warn(old_gpu_warn % (d, name, major, capability[1]))
 
+        total_mem = get_total_memory(d)
+        if total_mem < int(4e9):
```
Comment (apaszke) [inline torch/cuda/__init__.py]: We should warn about imbalance, not about not having top-of-the-shelf GPUs

File: torch/cuda/__init__.py [hunk 3]
```diff
@@ -287,6 +299,19 @@ def get_device_capability(device):
         return torch._C._cuda_getDeviceCapability(device)
 
 
+def get_total_memory(device):
```
Comment (apaszke) [inline torch/cuda/__init__.py]: I put that comment in another PR too. Can we just expose the device properties struct?

**Comment by lemairecarl** [comment]
>We should warn about imbalance, not about not having top-of-the-shelf GPUs

Yes, I agree. But could we find a reasonable threshold?

Here is the output from Tensorflow:
`Ignoring gpu device (device: 1, name: GeForce GT 710, pci bus id: 0000:03:00.0) with Cuda multiprocessor count: 1. The minimum required count is 8. You can adjust this requirement with the env var TF_MIN_GPU_MULTIPROCESSOR_COUNT.`

This seems reasonable to me. Most serious users will have a GPU with > 1 multiprocessor. If someone wants to use a weak GPU, they can edit the env var.

File: torch/cuda/__init__.py (hunk 2 above)
Comment (lemairecarl) [reply torch/cuda/__init__.py]: see my response below

**Comment by apaszke** [comment]
@lemairecarl I think we should take device properties and e.g. have a 25% threshold on imbalance in both memory and #CUDA cores. Also, don't check all GPUs, but those that are explicitly specified to be used in data parallel. If someone has 4 old GPUs then they can use DP, we should only warn that it runs as slow as the slowest one.

**Comment by lemairecarl** [comment]
@apaszke Makes sense; memory imbalance with DataParallel is the exact cause of my problem.

Though, outside of the DataParallel case, if my old GPU was at the index 0, would PyTorch try to use it? Or does it use the GPU with the largest memory?

**Comment by apaszke** [comment]
It will use it, but you can e.g. use `torch.cuda.set_device` to change the default one to a different index globally.

**Comment by lemairecarl** [comment]
Fine for me.

Could you point to an example of a struct being exposed? I'd like to do my first C++ contribution!

**Comment by apaszke** [comment]
@lemairecarl I think the simplest way would be to use `pybind` to define a Python class for `cudaDeviceProp`, and then expose the fields as members. You can grep for `py::class_` to see example bindings we already have in our sources, and `pybind` has a fairly good manual in case you need to use anything more complicated.

**Comment by vadimkantorov** [comment]
I've had an issue with old GPU preventing compilation: https://github.com/pytorch/pytorch/issues/4716 until hidden. I realize at install time we don't have PyTorch libs, but some sort of CMake warning / error based on parsing of nvidia-smi xml results would be helpful.

**Comment by lemairecarl** [comment]
@vadimkantorov That's strange. I can compile pytorch from source even with my old GPU visible.

File: torch/csrc/cuda/Module.cpp [hunk 4]
```diff
@@ -398,6 +399,13 @@ bool THCPModule_initCuda(PyObject *torch_module) {
   END_HANDLE_TH_ERRORS_RET(false)
 }
 
+cudaDeviceProp *getCudaDeviceProp(int device)
```
Comment (lemairecarl) [inline torch/csrc/cuda/Module.cpp]: Could also be a lambda, directly defined in `m.def("get_device_properties", ...)`

File: torch/nn/parallel/data_parallel.py [hunk 5]
```diff
@@ -60,9 +91,13 @@ def __init__(self, module, device_ids=None, output_device=None, dim=0):
         self.module = module
         self.device_ids = device_ids
         self.output_device = output_device
+        torch.cuda._lazy_call(self.check_cuda_balance)
```
Comment (lemairecarl) [inline torch/nn/parallel/data_parallel.py]: We could also use a lambda here.

```python
torch.cuda._lazy_call(lambda: _check_balance(self.device_ids))
```

File: torch/nn/parallel/data_parallel.py [hunk 6]
```diff
@@ -1,10 +1,41 @@
...
 from ..modules import Module
 from .scatter_gather import scatter_kwargs, gather
 from .replicate import replicate
 from .parallel_apply import parallel_apply
 
 
+def _check_balance(device_ids):
+    imbalance_warn = """
+    There is an imbalance between your GPUs. You may want to exclude GPU {} which
+    has >25% less memory or cores than GPU {}. You can do so by setting the
```
Comment (apaszke) [inline torch/nn/parallel/data_parallel.py]: ">25% less" is a bit hard to understand. How about "less than 75% of memory or cores ..."

File: torch/nn/parallel/data_parallel.py [hunk 7]
```diff
@@ -1,10 +1,41 @@
...
+            mem_min_i = i
+        if prop.total_memory > mem_max:
+            mem_max = prop.total_memory
+            mem_max_i = i
+        if prop.multi_processor_count < cores_min:
+            cores_min = prop.multi_processor_count
+            cores_min_i = i
+        if prop.multi_processor_count > cores_max:
+            cores_max = prop.multi_processor_count
+            cores_max_i = i
```
Comment (apaszke) [inline torch/nn/parallel/data_parallel.py]: a shorter way:
```python
def warn_imbalance(device_ids, dev_props, get_prop):
    values = [get_prop(props) for props in dev_props]
    min_pos, min_val = min(enumerate(values), key=operator.itemgetter(1))
    max_pos, max_val = max(enumerate(values), key=operator.itemgetter(1))
    if min_val / max_val < 0.75:
        warnings.warn(imbalance_warn.format(device_ids[min_pos], device_ids[max_pos]))
        return True
    return False

if warn_imbalance(lambda props: props.total_memory):
    return
if warn_imbalance(lambda props: props.multi_processor_count):
    return
```

**Comment by lemairecarl** [comment]
```
In [1]: from torch.nn.parallel import DataParallel

In [2]: from torch.nn.modules import Conv2d

In [3]: m = Conv2d(2, 2, 2)

In [4]: d = DataParallel(m)

In [5]: d = d.cuda()
/home/local/USHERBROOKE/lemc2220/source/pytorch/torch/nn/parallel/data_parallel.py:37: UserWarning: 
    There is an imbalance between your GPUs. You may want to exclude GPU 1 which
    has >25% less memory or cores than GPU 0. You can do so by setting the
    device_ids argument to DataParallel, or by setting the CUDA_VISIBLE_DEVICES
    environment variable.
  warnings.warn(imbalance_warn.format(weak_i, good_i))

In [6]: import torch.cuda as cu

In [7]: cu.get_device_properties(0)
Out[7]: GeForce GTX TITAN X | CUDA 5.2 | Memory: 12800 MB | Processors: 24

In [8]: cu.get_device_properties(1)
Out[8]: GeForce GT 710 | CUDA 3.5 | Memory: 1027 MB | Processors: 1
```

One thing to note is that `torch.cuda.get_device_properties()` does not exist before CUDA is initialized.

File: torch/nn/parallel/data_parallel.py (hunk 5 above)
Comment (apaszke) [reply torch/nn/parallel/data_parallel.py]: I don't think we have to use _lazy_call here. Just call `_check_balance(self.device_ids)` directly

File: torch/csrc/cuda/Module.cpp [hunk 8]
```diff
@@ -409,7 +417,16 @@ PyObject * THCPModule_initExtension(PyObject *self)
...
     return NULL;
   }
-  Py_RETURN_NONE;
+
+  auto m = py::handle(torch_module).cast<py::module>();
+  py::class_<cudaDeviceProp>(m,"CudaDeviceProperties")
+          .def_readwrite("major", &cudaDeviceProp::major)
+          .def_readwrite("minor", &cudaDeviceProp::minor)
+          .def_readwrite("multi_processor_count", &cudaDeviceProp::multiProcessorCount)
+          .def_readwrite("total_memory", &cudaDeviceProp::totalGlobalMem);
```
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/cuda/Module.cpp]: Also, can you also expose all the other properties (at least those that are easy to expose, like integer values)?

File: torch/csrc/cuda/Module.cpp [hunk 9]
```diff
@@ -409,7 +417,16 @@ PyObject * THCPModule_initExtension(PyObject *self)
   if (!THCPModule_initCuda(torch_module)) {
     return NULL;
   }
-  Py_RETURN_NONE;
+
+  auto m = py::handle(torch_module).cast<py::module>();
+  py::class_<cudaDeviceProp>(m,"CudaDeviceProperties")
```
Comment (apaszke) [inline torch/csrc/cuda/Module.cpp]: Can you rename this to `_CudaDeviceProperties`? We don't want people to ever think it is a public thing they can construct

**Review by apaszke — COMMENTED** [review]
Looks nice! I put a few suggestions inline. Thanks!

**Comment by apaszke** [comment]
Good point. Let's make it `_get_device_properties` in C++, and then have a Python function that calls it?

**Comment by lemairecarl** [comment]
Sure. But I couldn't find how to convert a `py::object` to a `PyObject *`...

**Comment by apaszke** [comment]
no, as I mentioned before this struct should be private, so it's enough to ensure that it is initialized before it's constructed for the first time.

**Comment by lemairecarl** [comment]
1. You didn't mention exactly that. Do you mean we'll have something like `torch._C._cuda_getDeviceProperties(device)` and no method added to `torch.cuda`?

2. I still can't get to return the `CudaDeviceProperties` other than by using pybind's `m.def(...)`. Here's what I've got:
    ```c
    PyObject * THCPModule_getDeviceProperties_wrap(PyObject *self, PyObject *arg)
    {
      ...

      std::unique_ptr<cudaDeviceProp> prop(new cudaDeviceProp);
      THCudaCheck(cudaGetDeviceProperties(prop.get(), device));
  
      py::object pybind_object = py::cast(prop);
      return pybind_object.ptr();  // get the PyObject *
      END_HANDLE_TH_ERRORS
    }
    ```

    which won't compile for unclear reasons:
    ```
    ...pybind11/cast.h:1614:66: error: cannot bind ‘const std::unique_ptr<cudaDeviceProp>’ lvalue to ‘std::unique_ptr<cudaDeviceProp>&&’
     return reinterpret_steal<object>(detail::make_caster<T>::cast(value, policy, parent));
    ```

**Comment by lemairecarl** [comment]
What works is to define `torch.cuda._get_device_properties()` with pybind and define a python method like this:
```python
def get_device_properties(device):
    if not _initialized:
        init();
    return _get_device_properties(device);
```

**Comment by apaszke** [comment]
Yes, that's exactly what I meant. But remember to change the string given to pybind when creating the struct wrapper to have an underscore at the beginning (so that people don't think they can construct it, as it would be exposed as `torch.cuda.CudaDeviceProperties`)

**Comment by lemairecarl** [comment]
Ah I see! Will do.

But one thing that bugs me is that `torch.cuda.get_device_names()` does not `init()` cuda, while `torch.cuda.get_device_properties()` will. Is this ok with you?

**Comment by lemairecarl** [comment]
>Also, can you also expose all the other properties (at least those that are easy to expose, like integer values)?

There's helluva bunch of fields in there (almost 100?). I don't think we should have all of those, especially those that are about textures.
http://docs.nvidia.com/cuda/cuda-runtime-api/structcudaDeviceProp.html

The ones that seem useful are:
* name
* major
* minor
* is_multi_gpu_board
* is_integrated
* multi_processor_count
* total_memory

and I now have included all of them,

**Comment by ngimel** [comment]
Wouldn't it be better to call THC_getCurrentDeviceProperties? It caches the properties, and cudaGetDeviceProperties is an expensive call.

**Comment by lemairecarl** [comment]
`torch.cuda.get_device_properties()` will not be used in a loop... I don't see why we should be concerned with speed here...

**Comment by apaszke** [comment]
Still, if we have this information cached, it's always better to use it

**Comment by lemairecarl** [comment]
I did not know the existence of `at::globalContext()` and `THCState::deviceProperties` (which caches the properties of all devices). Thanks for teaching me @ngimel ! I have many things to learn.

**Comment by lemairecarl** [comment]
All set!

@apaszke you may have missed these two comments of mine:
https://github.com/pytorch/pytorch/pull/5376#issuecomment-368635770
https://github.com/pytorch/pytorch/pull/5376#issuecomment-368639919

**Comment by ezyang** [comment]
@lemairecarl This is super late, but on the subject of pybind11 conversions, you may find https://github.com/pybind/pybind11/issues/1201 helpful

**Comment by ezyang** [comment]
@lemairecarl This is super late, but on the subject of pybind11 conversions, you may find https://github.com/pybind/pybind11/issues/1201 helpful

**Review by apaszke — COMMENTED** [review]
Just one minor thing. LGTM

File: torch/csrc/cuda/Module.cpp [hunk 10]
```diff
@@ -382,7 +383,28 @@ PyObject * THCPModule_initExtension(PyObject *self)
...
+    .def_readonly("name", &cudaDeviceProp::name)
+    .def_readonly("major", &cudaDeviceProp::major)
+    .def_readonly("minor", &cudaDeviceProp::minor)
+    .def_readonly("is_multi_gpu_board", &cudaDeviceProp::isMultiGpuBoard)
+    .def_readonly("is_integrated", &cudaDeviceProp::integrated)
+    .def_readonly("multi_processor_count", &cudaDeviceProp::multiProcessorCount)
+    .def_readonly("total_memory", &cudaDeviceProp::totalGlobalMem)
+    .def("__repr__", [](const cudaDeviceProp &prop) {
+      char repr[256];
+      snprintf(repr, 256, "_CudaDeviceProperties(name=\"%s\", major=%d, minor=%d, total_memory=%dMB, multi_processor_count=%d)",
```
Comment (apaszke) [inline torch/csrc/cuda/Module.cpp]: oh there's a much simpler way:
```cpp
std::ostringstream stream;
stream << "_CudaDeviceProperties(name=" << prop.name << ", major=" << prop.major
       << ", minor=" << prop.minor << ", total_memory=" << prop.totalGlobalMem / (1024 * 1024)
       << ", multi_processor_count=" << prop.multiProcessorCount << ")";
```

Comment (lemairecarl) [reply torch/csrc/cuda/Module.cpp]: ok, wasn't sure if we wanted to include `stringstream` only for that. will change it.

ah, I wish C++ had `"{}".format()`... I prefer keeping the format and the data apart!

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
Thanks a lot @lemairecarl!

**Comment by lemairecarl** [comment]
You're very welcome!
